### PR TITLE
Updates to TFM support for reference pkg src generator

### PR DESCRIPTION
**Pull request by mthalman**
In order to support packages with net6.0 TFM, we need to ensure that `DisableImplicitFrameworkReferences` is set to false (it's set to true by default) so that it compile against the existing targeting pack. There already existed code to set this to false but it was done globally for the entire project based on the presence of certain TFMs. But applying it for all TFMs is problematic because it can lead to build errors. For example, if applied to a `netstandard2.0` TFM, you can get this error:

```
 error NETSDK1023: A PackageReference for 'NETStandard.Library' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs
```

The solution is to conditionally apply the setting of the `DisableImplicitFrameworkReferences` property based on the target framework. I've made changes to incorporate this with an existing `PropertyGroup` that sets the output path.

One additional change was made to add `netcoreapp3.1` to the excluded TFMs.

Fixes https://github.com/dotnet/source-build/issues/3251

### CODE REVIEW

File: src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs [hunk 1]
```diff
@@ -40,7 +40,7 @@ public override bool Execute()
             string projectTemplateContent = File.ReadAllText(ProjectTemplate);
             string pkgProjectOutput = projectTemplateContent;
             string packageReferenceIncludes = "\n";
-            string outputPathByTfm = "\n";
+            string propertiesByTfm = "\n";
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs:43]: typo in the above suggestion. Use `tfmSpecificProperties` instead.

File: src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj [hunk 2]
```diff
@@ -7,7 +7,7 @@
     <PackagesTargetDirectory>$(RepoRoot)src\referencePackages\src\</PackagesTargetDirectory>
     <!-- The following target frameworks aren't buildable with the current SDK and as they don't contribute to the set of
          source build target verticals, can be safely excluded. -->
-    <ExcludeTargetFrameworks>netcore50;net463;portable-*;xamarin*;monotouch*;monoandroid*;win8;wp8;wpa81</ExcludeTargetFrameworks>
+    <ExcludeTargetFrameworks>netcoreapp3.1;netcore50;net463;portable-*;xamarin*;monotouch*;monoandroid*;win8;wp8;wpa81</ExcludeTargetFrameworks>
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj:10]: Why do we need to exclude netcoreapp3.1 if a targeting pack for it exists? https://github.com/dotnet/source-build-reference-packages/tree/main/src/targetPacks/ILsrc/microsoft.netcore.app.ref/3.1.0

File: src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs [hunk 3]
```diff
@@ -135,12 +135,19 @@ public override bool Execute()
...
                 {
-                    outputPathByTfm += $"  <PropertyGroup Condition=\" '$(TargetFramework)' == '{targetFramework}' \">\n";
-                    outputPathByTfm += $"    <OutputPath>$(ArtifactsBinDir){relativePath}/{subPath}/</OutputPath>\n";
-                    outputPathByTfm += $"  </PropertyGroup>\n\n";
+                    propertiesByTfm += $"    <OutputPath>$(ArtifactsBinDir){relativePath}/{subPath}/</OutputPath>\n";
                 }
+                if (targetFramework == "netstandard2.0" ||
+                    targetFramework == "netstandard2.1" ||
+                    targetFramework == "netcoreapp3.0" ||
+                    targetFramework == "net6.0")
```
Comment (MichaelSimons) [inline src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs:146]: Are we going to need this for all versions going forward?  If so, can we make this check more abstract (e.g. netx.y)?

File: src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj:10]: I don't know the history here but perhaps it was because of the underlying issue being fixed here.  That being said, I don't suspect we will need any netcoreapp3.1 TFMs going forward.  I like the idea of filtering the TFM until it is proven they are needed.

File: src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs:146]: I want to send a follow-up PR anyway that cleans this up. My idea is that we shouldn't disable implicit framework references at all and just make sure that the right packages are picked up (for all target frameworks). I'm fine with this change and think we should take it as is, as it unblocks targeting net6.0.

File: src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj (hunk 2 above)
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj:10]: Based on that I would that think that we should block all netcoreapp* TFMs and remove their targeting packs. But if we would then stumble upon a package that has one of these TFMs in it and we actually require it, we would need to bring the targeting pack back and remove the exclusion filter. That just makes me wonder why do we want to block netcoreapp3.1 but not netcoreapp3.0, netcoreapp2.2, netcoreapp2.1 or netcoreapp2.0?

File: src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs [hunk 4]
```diff
@@ -135,12 +135,19 @@ public override bool Execute()
...
 
+                propertiesByTfm += $"  <PropertyGroup Condition=\" '$(TargetFramework)' == '{targetFramework}' \">\n";
                 if (subPath == "lib")
                 {
-                    outputPathByTfm += $"  <PropertyGroup Condition=\" '$(TargetFramework)' == '{targetFramework}' \">\n";
-                    outputPathByTfm += $"    <OutputPath>$(ArtifactsBinDir){relativePath}/{subPath}/</OutputPath>\n";
-                    outputPathByTfm += $"  </PropertyGroup>\n\n";
+                    propertiesByTfm += $"    <OutputPath>$(ArtifactsBinDir){relativePath}/{subPath}/</OutputPath>\n";
                 }
+                if (targetFramework == "netstandard2.0" ||
```
Comment (ViktorHofer) [inline src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs:143]: netstandard2.0 doesn't use implicit framework references. Why do you need to set `DisableImplicitFrameworkReferences` to false for it? I thought that you mentioned in your PR summary that you can't globally set the property to false for all TFMs exactly because of netstandard2.0.

File: src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj:10]: Let's take a step back and look at this holistically and establish a set of principles.  We can log a follow-up issue(s) to achieve these as this is well beyond the scope of what Matt is trying to accomplish here - unblock the generation of net 6.0 SBRPs.

**Principles**
1. The SBRPs in this repo should contain the minimum set of TFMs required by source-build and the product repos.  In the past, the SBRPs contained whatever TFMs the tooling emitted.  Very seldom did folks specify the explicit minimal set of TFMs required by the product.
1. The tooling should support all TFMs required by source-build and the product repos. We know there are always repos that lag behind in what TFMs they target.  If an SBRP is needed for this type of situation, the tooling should be capable of generating ref packs for those TFMs.  Being capable is very different than supporting it "out of the box".

To help achieve the first principal, I propose the tooling filters out a default set of TFMs.  This filter should be ignored when an explicit set of TFMs is specified.  Furthermore, we can create some tooling or modify the prebuilt reporting infra to indicate which TFMs are consumed from the detected prebuilts.  If agreed upon we can log an issue for these and debate what the default TFM filters should be.

File: src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs (hunk 4 above)
Comment (mthalman) [reply src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs:143]: I did mention that, didn't I? 🫤

So it looks like `netstandard2.0` TFM for Microsoft.Bcl.AsyncInterfaces.6.0.0 will result in the following error if `DisableImplicitFrameworkReferences` is `true`:

```
error CS0234: The type or namespace name 'CompilationRelaxationsAttribute' does not exist in the namespace 'System.Runtime.CompilerServices' (are you missing an assembly reference?)
```

Setting `DisableImplicitFrameworkReferences` to false will make those errors go away.

However, for `netstandard2.0` of System.Runtime.CompilerServices.Unsafe.6.0.0, it will result in the following error if `DisableImplicitFrameworkReferences` is `false` (the opposite of the previous example):

```
error NETSDK1023: A PackageReference for 'NETStandard.Library' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs
```

So in reality, there doesn't seem to be one solution that solves both without manual fixup.

File: src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj (hunk 2 above)
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj:10]: Thanks for sharing your thoughts on this.

> To help achieve the first principal, I propose the tooling filters out a default set of TFMs. This filter should be ignored when an explicit set of TFMs is specified.

Let me describe how the current include / exclude mechanism works. Target frameworks that are excluded via the [`ExcludedTargetFrameworks`](https://github.com/dotnet/source-build-reference-packages/blob/a848f811a52d75494029e663cd0a799be11744fe/src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj#L10) property are always excluded, even if they are explicitly listed in the [`IncludeTargetFrameworks`](https://github.com/dotnet/source-build-reference-packages/blob/a848f811a52d75494029e663cd0a799be11744fe/src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj#L71) property. Reason for that is that those TFMs aren't restore or build-able with the SDK used (either because of missing targeting packs or because the SDK just doesn't support targeting the TFM).

If a user invokes the following command `generate.sh --package "System.Drawing.Common,7.0.0,net7.0"` then that means that the `IncludeTargetFrameworks` property is set to `net7.0` and only the net7.0 TFM will be generated.

Let's assume that another package `System.X` with version `x.x.x` contains a `netcore50` compile assembly. If a user invokes `generate.sh --package "System.X,x.x.x,net7.0;netcore50"`, then only the `net7.0` TFM is generated as `netcore50` is excluded as it isn't restore and/or buildable.

Long story short, we are conflating reasons for excluding a TFM. If we want to (by default) exclude TFMs that are supported by the SDK, then we should add the following snippet instead:

```suggestion
    <ExcludeTargetFrameworks>netcore50;net463;portable-*;xamarin*;monotouch*;monoandroid*;win8;wp8;wpa81</ExcludeTargetFrameworks>
    <!-- The following target frameworks should be excluded even though they are supported by the current SDK because there is no need to have them.
         Only exclude them when target frameworks to include aren't provided. -->
    <ExcludeTargetFrameworks Condition="'$(IncludeTargetFrameworks)' == ''">$(ExcludeTargetFrameworks);netcoreapp3.1</ExcludeTargetFrameworks>
```

File: src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs (hunk 4 above)
Comment (ViktorHofer) [reply src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs:143]: To me that indicates that we are missing some details here.

> Setting DisableImplicitFrameworkReferences to false will make those errors go away.

That's because you didn't to update this block: https://github.com/dotnet/source-build-reference-packages/blob/a848f811a52d75494029e663cd0a799be11744fe/src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs#L58-L61

 Remove the `&& !includesNetStandard21 && !includesNetCoreApp30` part in the condition as we now define the `DisableImplicitFrameworkReferences` property per TFM and don't need to conflate these together anymore.

File: src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj (hunk 2 above)
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj:10]: Thanks for filing. Still, for this PR I think @mthalman should take the above suggestion.

[the remark this answers is not in the document]
Comment (mthalman) [reply src/referencePackageSourceGenerator/ReferencePackageSourceGenerator.proj]: I've applied the changes suggested in https://github.com/dotnet/source-build-reference-packages/pull/567#discussion_r1132454025.

File: src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs (hunk 4 above)
Comment (mthalman) [reply src/referencePackageSourceGenerator/ReferencePackageSourceTask/GenerateProject.cs]: Thanks. That fixes it.

**Review by MichaelSimons — APPROVED** [review]

**Comment by mthalman** [comment]
Sorry, @ViktorHofer. I shouldn't have enabled auto-merge to give everyone a chance to review. Let me know if further changes are needed.

**Comment by ViktorHofer** [comment]
Looks good 👍
